### PR TITLE
Add models for loan usecase

### PR DESCRIPTION
**Pull request by adelowo**

### CODE REVIEW

File: intents.go [hunk 1]
```diff
@@ -46,6 +46,23 @@ type Stake struct {
...
+
+type WithdrawSupply struct {
+	Type     AssetType `json:"type,omitempty" binding:"required"`
+	Currency string    `json:"currency,omitempty" binding:"required"`
+	Amount   string    `json:"amount,omitempty" binding:"required"`
+
+	// Contract address of the project to withdraw from
+	Address string `json:"address,omitempty" binding:"required"`
+}
+
```
Comment (blewater) [inline intents.go]: Example values in comments, e.g., protocol,

File: intents.go [hunk 2]
```diff
@@ -46,6 +46,23 @@ type Stake struct {
...
+	Type    AssetType `json:"type" binding:"required"`
+	Address string    `json:"address" binding:"required"`
+
+	// can be empty? and solver chooses a default protocol to supply to?
+	Protocol string `json:"protocol,omitempty"`
+}
+
+type WithdrawSupply struct {
+	Type     AssetType `json:"type,omitempty" binding:"required"`
+	Currency string    `json:"currency,omitempty" binding:"required"`
```
Comment (blewater) [inline intents.go]: Asset?

**Review by blewater — CHANGES_REQUESTED** [review]

File: intents.go [hunk 3]
```diff
@@ -44,6 +46,33 @@ type Stake struct {
...
+	Protocol string `json:"protocol,omitempty"`
+}
+
+type WithdrawSupply struct {
+	Type AssetType `json:"type,omitempty" binding:"required"`
+	// Asset is the contract address for the token to withdraw or add into the
+	// Protocol
+	Asset string `json:"asset,omitempty" binding:"required"`
+
+	Amount string `json:"amount,omitempty" binding:"required"`
```
Comment (blewater) [inline intents.go]: Why string for amount? I see that we previously used float64, which is limited, but float64 should be the choice for consistency. Ideally, we can also change them all to decimals as long as we coordinate the change with the codebases, i.e., Solver.
1. Strings don't enforce any constraints on the format or precision of the numeric value.
2. It is error-prone. We'd need to parse the strings into actual numeric types first.
3. less efficient in terms of memory usage and performance than using a proper numeric type.

File: intents.go [hunk 4]
```diff
@@ -44,6 +46,33 @@ type Stake struct {
...
+	Type AssetType `json:"type,omitempty" binding:"required"`
+
+	// Asset is the contract address for the token to Supply or add into the
+	// Protocol
+	Asset string `json:"asset,omitempty" binding:"required"`
+
+	// can be empty? and solver chooses a default protocol to supply to?
+	// this would be the Contract address for the Protocol
+	// Aave3 as an example would be 0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2
+	Protocol string `json:"protocol,omitempty"`
```
Comment (blewater) [inline intents.go]: Now i understand that it is a contract address? Why not a Geth address type? Also the name should indicate that.

File: intents.go [hunk 5]
```diff
@@ -44,6 +46,33 @@ type Stake struct {
 	ChainId string    `json:"chainId"`
 }
 
+type Supply struct {
```
Comment (blewater) [inline intents.go]: Could at least a unit test be added?

**Review by blewater — CHANGES_REQUESTED** [review]

File: intents.go (hunk 3 above)
Comment (adelowo) [reply intents.go]: I was following existing types we use for addresses and others. I would change that.

File: intents.go (hunk 4 above)
Comment (adelowo) [reply intents.go]: I was following existing types we use for addresses and others. I would change that.

Comment (blewater) [reply intents.go]: I just noticed this [Notion Loan-and-Leverage Use case](https://www.notion.so/balloondogs/Loan-and-Leverage-82c61a363cce4a85974adfb971bca50f?pvs=4)
Please use that naming convention closely and link the documentation to the PR. For example, currency is the token address, and it's using string.

Comment (blewater) [reply intents.go]: you're right about the address data type. We're using a string data type in the model.

File: intents.go [hunk 6]
```diff
@@ -14,8 +14,10 @@ import (
 type AssetType string
 
 const (
-	TokenType AssetType = "TOKEN"
-	StakeType AssetType = "STAKE"
+	TokenType          AssetType = "TOKEN"
+	StakeType          AssetType = "STAKE"
+	SupplyType         AssetType = "SUPPLY"
+	WithdrawSupplyType AssetType = "WITHDRAW_SUPPLY"
 )
```
Comment (ronezra3) [inline intents.go:24]: AssetType = "WITHDRAW"

Comment (adelowo) [reply intents.go:24]: done in https://github.com/blndgs/model/pull/23/commits/a4180516f38df4f16244c106c6668df00c85fa10

File: intents.go [hunk 7]
```diff
@@ -19,10 +19,9 @@ import (
...
 
 const (
-	TokenType          AssetType = "TOKEN"
-	StakeType          AssetType = "STAKE"
-	SupplyType         AssetType = "SUPPLY"
-	WithdrawSupplyType AssetType = "WITHDRAW"
+	TokenType AssetType = "TOKEN"
+	StakeType AssetType = "STAKE"
+	LoanType  AssetType = "LOAN"
 )
```
Comment (ronezra3) [inline intents.go:24]: @singhhp1069 No need for more generic approach?

File: intents.go [hunk 8]
```diff
@@ -321,12 +320,27 @@ func unmarshalTransactional(data json.RawMessage) (Transactional, error) {
...
-		var withdraw WithdrawSupply
+	case LoanType:
+		// we use the LoanType for both LoanTo and LoanFrom
+		// the only difference is the Amount field present in LoanFrom,
+		// we can use that to determine the struct to unmarshal into
+
+		var typeDetectStruct struct {
+			Amount string `json:"amount"`
+		}
+
```
Comment (ronezra3) [inline intents.go]: what typeDetectStruct used for? Maybe a better name?

File: intents_test.go [hunk 9]
```diff
@@ -269,34 +269,34 @@ func TestSubmitHandler(t *testing.T) {
...
-							Currency: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
-							Amount:   "104",
+						From: LoanFrom{
+							Type:    LoanType,
+							Address: "0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2",
+							Asset:   "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
+							Amount:  "104",
 						},
 						To: Asset{
 							Type:    TokenType,
```
Comment (ronezra3) [inline intents_test.go:302]: What about Approve pre Supply?

**Review by ronezra3 — COMMENTED** [review]
@singhhp1069 @adelowo

File: intents_test.go (hunk 9 above)
Comment (adelowo) [reply intents_test.go:302]: this is just a test to make sure we can validate the contract addresses and amount

File: intents.go (hunk 8 above)
Comment (adelowo) [reply intents.go]: no longer needed since we merged both types into one

**Comment by adelowo** [comment]
closing this because of #24

**Pull request closed without merging**